File: Project Enhancements/Databases/src/src/Appointment.java
package src;

import java.util.Date;
import java.time.LocalDate;

public class Appointment {
	private final String appID;
	private Date date;
	private String desc;
	
	/**
	 * @param appID
	 * @param date
	 * @param desc
	 */
	public Appointment(String appID, Date date, String desc) {
		if (appID == null || appID.length() > 10 || appID.isEmpty()) {
			throw new IllegalArgumentException("Invalid ID");
		}
		if (date == null || date.before(new Date())) {
			throw new IllegalArgumentException("Invalid Date");
		}
		if (desc == null || desc.length() > 50 || desc.isEmpty()) {
			throw new IllegalArgumentException("Invalid Description");
		}
		
		
		this.appID = appID;
		this.date = date;
		this.desc = desc;
	}

	/**
	 * @return the date
	 */
	public Date getDate() {
		return date;
	}

	/**
	 * @param date the date to set
	 */
	public void setDate(Date date) {
		if (date == null || date.before(new Date())) {
			throw new IllegalArgumentException("Invalid Date");
		}
		
		this.date = date;
	}

	/**
	 * @return the desc
	 */
	public String getDesc() {
		
		return desc;
	}

	/**
	 * @param desc the desc to set
	 */
	public void setDesc(String desc) {
		if (desc == null || desc.length() > 50 || desc.isEmpty()) {
			throw new IllegalArgumentException("Invalid Description");
		}
		
		this.desc = desc;
	}

	/**
	 * @return the appID
	 */
	public String getAppID() {
		return appID;
	}
	
	
	
}

File: Project Enhancements/Software Design & Engineering/Brea_SoftDesignEngineeringEnhancement/Source/src/Appointment.cpp
#include "Appointment.h"
#include <ctime>
#include <stdexcept>

Appointment::Appointment(const std::string& appID, std::time_t date, const std::string& desc) 
    : appID(appID), date(date), desc(desc) {

    //validate ID
    if (appID.empty() || appID.length() > 10) {
        throw std::invalid_argument("Invalid ID");
    }

    //validate other member variables
    validateDate(date);
    validateDesc(desc);
}

//return appointment ID
std::string Appointment::getAppID() const {
    return appID;
}

//return appointment date
std::time_t Appointment::getDate() const {
    return date;
}

//return appointment description
std::string Appointment::getDesc() const {
    return desc;
}

//set new date for appointment
void Appointment::setDate(std::time_t date) {
    validateDate(date);
    this->date = date;
}

//set new description for appointment
void Appointment::setDesc(const std::string& desc) {
    validateDesc(desc);
    this->desc = desc;
}

//makes sure date meets required input guidelines
void Appointment::validateDate(std::time_t date) {
    if (date < std::time(nullptr)) {
        throw std::invalid_argument("Invalid Date");
    }
}

//makes sure description meets required input guidelines
void Appointment::validateDesc(const std::string& desc) {
    if (desc.empty() || desc.length() > 50) {
        throw std::invalid_argument("Invalid Description");
    }
}